tests/instrument/expr_match.ml.reference

```diff
@@ -99,3 +99,174 @@ let f x =
         (match x with
          | Foo  -> (___bisect_mark___expr_match 36; print_endline "foobar")
          | Bar  -> (___bisect_mark___expr_match 37; print_endline "barbar"))))
+let f x =
+  ___bisect_mark___expr_match 43;
+  (match x with
+   | Foo |Bar  as ___bisect_matched_value___ ->
+       ((((match ___bisect_matched_value___ with
+           | Foo  -> (___bisect_mark___expr_match 41; ())
+           | Bar  -> (___bisect_mark___expr_match 42; ())
+           | _ -> ()))[@ocaml.warning "-8-11"]);
+        print_endline "foo"))
```

Comment:
Yes, this was my first instinct. The main reason against doing "full promotion" is that if there is a bug in Bisect_ppx's pattern rewriter, it can mess up the meaning of the user's code. That could silently make their testing misleading or useless.

With nested match expressions like now, the bug might mess up coverage statistics, but it should not harm the correctness of code under test. There are other, lamer, reasons, such as it is easier to wrap the pattern with `as ___bisect_matched_value___` than to move the case body out to a let-expression, find all its free variables, turn it into a function, and then apply it. That's what I think we would have to do, and, by the way, it might also look strange, at first, if someone examined the output.

As for power users, I think we will have to be clear that the advantage of repeated matching, as proposed now, is correctness of testing, and that the disadvantage only applies during testing and has nothing to do with their production code anyway. Not sure how to go about this, maybe a FAQ will be needed, or maybe it will be a non-problem anyway.

In any case, the algorithm I have now, and the algorithm you are asking about, are near relatives of each other, and I wrote the `transform_pattern` function in such a way that it can be used to do what you want instead. It should be only slightly annoying to change between the two, if it turns out to be necessary or desirable.

Let me know your thoughts.